konan_sdk/endpoints/base_endpoint.py

```diff
@@ -0,0 +1,97 @@
+import json
+import requests
+from loguru import logger
+from typing import Dict, Union
+from abc import abstractmethod
+
+
+class KonanBaseEndpoint:
+    """
+    Base Endpoint class for Konan endpoints to inherit.
+    """
+
+    name = ''  # endpoint name to reference in logs
+
+    def __init__(self, api_url: str, user, **kwargs) -> None:
+
+        self.api_url = api_url
+        self.user = user
+
+        # request headers
+        self.headers = {}
+        # response object
+        self.response = {}
+
+    @property
+    def request_url(self):
+        return self.api_url + self.endpoint_path
+
+    @property
+    @abstractmethod
+    def endpoint_path(self):
+        # Override method to return path of specified endpoint
+        pass
+
+    def prepare_request(self):
+        # Override method with logic to implement before sending request
+        pass
+
+    def process_response(self):
+        # Override method with logic to implement after receiving response
+        pass
+
+    def post(self, payload: Union[Dict, str]):
+        """
+        Base POST function for all post endpoints
+        """
+        self.prepare_request()
+
+        # If request content type is json, convert input dict to json string
+        if self.headers and dict(self.headers).get('Content-Type', '') == 'application/json' and isinstance(payload, Dict):
+            payload = json.dumps(payload)
+
+        logger.debug(f"Sending {self.name} request")
+
+        # send request
+        response = requests.post(
+            self.request_url,
+            headers=self.headers,
```

Comment:
I might be missing something but I don't see the `self.headers` being set except in `__init__` , I believe we should check first if the user is not null and if its not null then we check if the headers contain the authorization header but I propose instead of all that, we directly set the headers in the `__init__` method to contain the auth header if the user is present instead of checking everytime in `get` and `post` like `PredictionEndpoint`